# EDA

Para describir la data que tenemos empecemos por la cantidad de datos que tenemos y las tesis:

## Formato

El formato de las tesis es archivos pdfs, al principio se determina los registros de las tesis que deben ser descargados, estos archivos fueron obtenidos desde el siguiente link: http://saber.ucv.ve/handle/10872/2/browse?type=dateissued&sort_by=2&order=ASC&rpp=8211&etal=0&submit_browse=Actualizar

En este link fue la manera en que obtenemos todas las tesis, dependiendo de la fecha de consulta puede que alla mas tesis subidas en la pagina. Para obtener este link se navega desde la pagina de inicio luego en 2) Tesis, luego se selecionar pregrado en el menu dropdown y luego en ir; se selecciona pregrado en las letras azules; esto nos devuelve al menu anterior pero con pregrado preseleccionado (el url donde se empieza es importante ya que si se selecciona el siguiente paso de Fecha de publicacion, siempre va a Tesis); Se escoge fecha de publicacion y nos da el siguiente url: http://saber.ucv.ve/handle/10872/2/browse?type=dateissued&submit_browse=Fecha+Publicaci%C3%B3n 

Si luego le damos al boton de actualizar para mostrar mas resultados nos damos cuenta que el valor `rpp` si es cambiado nos pueden enseñar todas las tesis que necesitamos el resultado es http://saber.ucv.ve/handle/10872/2/browse?type=dateissued&sort_by=2&order=ASC&rpp=30&etal=0&submit_browse=Actualizar

De esta forma se logro obtener el link para descargar todas las tesis en una sola pagina de HTML.

### Datos a extraer

Ahora, los datos a extraer de las tesis son: 
- Año de la tesis (Ejm: 27-May-2011)
- Titulo de la tesis (Ejm: Estudio de sobretensiones por maniobra debido a energización de bancos de condensadores y transformadores en líneas de transmisión de 115kV) 
- Autor de la tesis (Ejm: De Gregorio S., Luis T.)
- Tamaño del archivo (Ejm: 6.77 MB)
- Link de la tesis (Ejm: http://saber.ucv.ve/handle/10872/11222)
- Link de direccion de tesis donde esta el pdf (este es el descargable) (Ejm: http://saber.ucv.ve/bitstream/10872/17712/1/TESIS%20DE%20GRADO%20LUIS%20DE%20GREGORIO.pdf)
- Resumen (Este dato esta vacio por la cantidad de erroes que se podian presentar, ya sea de caracteres o de falta de resumen)

### Para extraer todos estos datos

Al final estos datos seran almacenados un un csv, que luego puede utilizarce este data frame para realizar la operacion de descarga de el link donde esta el pdf, los demas datos pueden ser utilizados como metada del archivo.

El archivo `scraperv3.py` es el encargado de obtener todos los datos de la tesis y guardalos en un archivo csv. El la funcion main se especifica la cantida de archivos a guardar, en este caso se utilizaron **8211**

Estos datos se guardaron en `csv_files/url_thesis_8211.csv` Estos fueron todos los datos


In [2]:
## Muestra de el archivo inicial con scrapers
import pandas as pd
import numpy as np

df = pd.read_csv("../csv_files/url_thesis_8211.csv")
df

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1         2         2010  La Economía del Ocio como una alternativa de e...   
2         3  27-May-2011  La producción de conocimiento : un análisis mé...   
3         4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4         5  27-May-2011  Evaluación geofísica integral de los conos alu...   
...     ...          ...                                                ...   
8206   8207   4-Dec-2020  Elaboración de un manual de procedimientos e i...   
8207   8208   8-Mar-2021  Condiciones de vida de los estudiantes de Geoq...   
8208   8209   8-Mar-2021  Análisis de los elementos pospornográficos de ...   
8209   8210  11-Mar-2021  La Fecundidad Adolescente en Venezuela vista a...   
8210   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
1                                     Rivera, Elizabeth       2 MB   
2     Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  995.92 kB   
3                                    D'Amico, Walter F.  506.32 kB   
4                  Cragno, Analia V.; Romero, Marcos A.    3.02 MB   
...                                                 ...        ...   
8206                               Armas R., Rodolfo J.  924.37 kB   
8207                     Pérez, Andrea; Pinto, Hendrith    7.21 MB   
8208                                   Belandria, Sofía    1.13 MB   
8209                              Simao, Edith Carolina    3.56 MB   
8210                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1     http://saber.ucv.ve/handle/10872/16245   
2        http://saber.ucv.ve/handle/10872/29   
3        http://saber.ucv.ve/handle/10872/32   
4        http://saber.ucv.ve/handle/10872/31   
...                                      ...   
8206  http://saber.ucv.ve/handle/10872/20906   
8207  http://saber.ucv.ve/handle/10872/20969   
8208  http://saber.ucv.ve/handle/10872/20970   
8209  http://saber.ucv.ve/handle/10872/20973   
8210  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN  
1     http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN  
2     http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN  
3     http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN  
4     http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN  
...                                                 ...      ...  
8206  http://saber.ucv.ve/bitstream/10872/20906/1/TE...      NaN  
8207  http://saber.ucv.ve/bitstream/10872/20969/1/CO...      NaN  
8208  http://saber.ucv.ve/bitstream/10872/20970/1/So...      NaN  
8209  http://saber.ucv.ve/bitstream/10872/20973/1/TF...      NaN  
8210  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN  

[8211 rows x 8 columns]

## Primer data frame: Todas las tesis

Este tiene 8211 filas y 8 columnas, de aqui se puede sacar los siguientes datos. Tamaño promedio de las tesis, año con más tesis.

El formato de el tamaño de las tesis es string este debe ser convertido a bytes y a flotante

In [3]:
size_df = df[['index', 'size']]
size_df

index       size
0         1    6.77 MB
1         2       2 MB
2         3  995.92 kB
3         4  506.32 kB
4         5    3.02 MB
...     ...        ...
8206   8207  924.37 kB
8207   8208    7.21 MB
8208   8209    1.13 MB
8209   8210    3.56 MB
8210   8211    2.33 MB

[8211 rows x 2 columns]

In [4]:
def get_bytes(size):
    # Para convertir los datos a bytes y a flotante
    number, unit = size.split()
    if unit == 'MB':
        number = float(number)
    if unit == 'kB':
        number = float(number) / 1000
    return number

In [5]:
df_byte = df.copy()

In [6]:
do_bytes = np.vectorize(get_bytes)

In [7]:
df_byte['size'] = do_bytes(df_byte['size'])

In [8]:
df_byte

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1         2         2010  La Economía del Ocio como una alternativa de e...   
2         3  27-May-2011  La producción de conocimiento : un análisis mé...   
3         4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4         5  27-May-2011  Evaluación geofísica integral de los conos alu...   
...     ...          ...                                                ...   
8206   8207   4-Dec-2020  Elaboración de un manual de procedimientos e i...   
8207   8208   8-Mar-2021  Condiciones de vida de los estudiantes de Geoq...   
8208   8209   8-Mar-2021  Análisis de los elementos pospornográficos de ...   
8209   8210  11-Mar-2021  La Fecundidad Adolescente en Venezuela vista a...   
8210   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author     size  \
0                               De Gregorio S., Luis T.  6.77000   
1                                     Rivera, Elizabeth  2.00000   
2     Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  0.99592   
3                                    D'Amico, Walter F.  0.50632   
4                  Cragno, Analia V.; Romero, Marcos A.  3.02000   
...                                                 ...      ...   
8206                               Armas R., Rodolfo J.  0.92437   
8207                     Pérez, Andrea; Pinto, Hendrith  7.21000   
8208                                   Belandria, Sofía  1.13000   
8209                              Simao, Edith Carolina  3.56000   
8210                        Utrera Ávila, Joselin Marly  2.33000   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1     http://saber.ucv.ve/handle/10872/16245   
2        http://saber.ucv.ve/handle/10872/29   
3        http://saber.ucv.ve/handle/10872/32   
4        http://saber.ucv.ve/handle/10872/31   
...                                      ...   
8206  http://saber.ucv.ve/handle/10872/20906   
8207  http://saber.ucv.ve/handle/10872/20969   
8208  http://saber.ucv.ve/handle/10872/20970   
8209  http://saber.ucv.ve/handle/10872/20973   
8210  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN  
1     http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN  
2     http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN  
3     http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN  
4     http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN  
...                                                 ...      ...  
8206  http://saber.ucv.ve/bitstream/10872/20906/1/TE...      NaN  
8207  http://saber.ucv.ve/bitstream/10872/20969/1/CO...      NaN  
8208  http://saber.ucv.ve/bitstream/10872/20970/1/So...      NaN  
8209  http://saber.ucv.ve/bitstream/10872/20973/1/TF...      NaN  
8210  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN  

[8211 rows x 8 columns]

In [9]:
df_byte.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8211 entries, 0 to 8210
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          8211 non-null   int64  
 1   thesis_year    8211 non-null   object 
 2   thesis_title   8211 non-null   object 
 3   thesis_author  8211 non-null   object 
 4   size           8211 non-null   float64
 5   thesis_link    8211 non-null   object 
 6   pdf_link       8211 non-null   object 
 7   resumen        0 non-null      float64
dtypes: float64(2), int64(1), object(5)
memory usage: 513.3+ KB


In [10]:
df_byte.describe()

index         size  resumen
count  8211.000000  8211.000000      0.0
mean   4106.000000     5.455386      NaN
std    2370.455863    14.826478      NaN
min       1.000000     0.003370      NaN
25%    2053.500000     1.360000      NaN
50%    4106.000000     2.670000      NaN
75%    6158.500000     5.150000      NaN
max    8211.000000   463.160000      NaN

In [11]:
size_GB = df_byte['size'].sum()/1000
size_GB

44.79417043000001

### Tamaño de las tesis
De estos valores la media es 5 MB con una desviacion estandar de 14.8 MB donde el valor minimo es 0.00337 MB o lo que es igual a 3.37 kB. El archivo más pesado siendo de 463 MB.

El tamaño total de tesis es 44.7 GB.

## Fecha más comun de las tesis

In [12]:
date_df = df_byte.copy()

In [13]:
date_df.head()

index  thesis_year                                       thesis_title  \
0      1         2009  Estudio de sobretensiones por maniobra debido ...   
1      2         2010  La Economía del Ocio como una alternativa de e...   
2      3  27-May-2011  La producción de conocimiento : un análisis mé...   
3      4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4      5  27-May-2011  Evaluación geofísica integral de los conos alu...   

                                       thesis_author     size  \
0                            De Gregorio S., Luis T.  6.77000   
1                                  Rivera, Elizabeth  2.00000   
2  Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  0.99592   
3                                 D'Amico, Walter F.  0.50632   
4               Cragno, Analia V.; Romero, Marcos A.  3.02000   

                              thesis_link  \
0  http://saber.ucv.ve/handle/10872/17712   
1  http://saber.ucv.ve/handle/10872/16245   
2     http://saber.ucv.ve/handle/10872/29   
3     http://saber.ucv.ve/handle/10872/32   
4     http://saber.ucv.ve/handle/10872/31   

                                            pdf_link  resumen  
0  http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN  
1  http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN  
2  http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN  
3  http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN  
4  http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN

In [14]:
date_df['thesis_year_dt'] = pd.to_datetime(date_df['thesis_year'])
date_df.head()

index  thesis_year                                       thesis_title  \
0      1         2009  Estudio de sobretensiones por maniobra debido ...   
1      2         2010  La Economía del Ocio como una alternativa de e...   
2      3  27-May-2011  La producción de conocimiento : un análisis mé...   
3      4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4      5  27-May-2011  Evaluación geofísica integral de los conos alu...   

                                       thesis_author     size  \
0                            De Gregorio S., Luis T.  6.77000   
1                                  Rivera, Elizabeth  2.00000   
2  Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  0.99592   
3                                 D'Amico, Walter F.  0.50632   
4               Cragno, Analia V.; Romero, Marcos A.  3.02000   

                              thesis_link  \
0  http://saber.ucv.ve/handle/10872/17712   
1  http://saber.ucv.ve/handle/10872/16245   
2     http://saber.ucv.ve/handle/10872/29   
3     http://saber.ucv.ve/handle/10872/32   
4     http://saber.ucv.ve/handle/10872/31   

                                            pdf_link  resumen thesis_year_dt  
0  http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN     2009-01-01  
1  http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN     2010-01-01  
2  http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN     2011-05-27  
3  http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN     2011-05-27  
4  http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN     2011-05-27

In [26]:
dates = date_df.groupby(date_df['thesis_year_dt'].dt.year)
dates.size()

thesis_year_dt
2009       1
2010       1
2011     594
2012     722
2013     709
2014     915
2015    1188
2016     504
2017    1817
2018     959
2019     621
2020     176
2021       4
dtype: int64

### Descripcion

Para la fecha de extración de 02 de Abril donde se encontraban 8211 tesis de pregrado, se observa la siguiente distribuccion de años:

- El año con mas tesis es 2017 con 1817 tesis
- El año con menos tesis son 2009 y 2010.


## Segundo data frame: Descargado o no?

Luego de tener cuales tesis descargar y el link de descarga de los pdfs se procede a descargar todas las tesis. Esta descarga puede dar diferentes tipos de errores, se procede a tener varios reintentos, con 10 segundos de espera con `backoff_factor` de 1 para cada reintento. Se hacen 5 reintentos:

```python
import requests
from requests.adapters import HTTPAdapter, Retry

s = requests.Session()
Retry(total=5, backoff_factor=1, status_forcelist=[ 502, 503, 504 ])
pdf_response = s.get(url, timeout=10.0)
```

Con 1 segundo de `backoff_factor` se realizan `total=5` reintentos entonces:
Se realizan reintentos en segundos:
[0.5, 1, 2, 4, 8]

El `timeout=10.0` en request es el tiempo que tarda esperando una respuesta antes de seguir.

`status_forcelist` Son los codigos de error donde se reintenta otra vez.

Luego de descargarlas se realiza una comparacion si el archivo existe, es decir si fue descargado. 

Esta comparacion se realiza en el archivo `scripts/preprocessing_files/check_existance.py` Este nos da un nuevo data frame.


In [38]:
## Muestra el archivo con los archivos descargados en una nueva columna
import pandas as pd
import numpy as np

df = pd.read_csv("../csv_files/url_thesis_8211_exist.csv")
df

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1         2         2010  La Economía del Ocio como una alternativa de e...   
2         3  27-May-2011  La producción de conocimiento : un análisis mé...   
3         4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4         5  27-May-2011  Evaluación geofísica integral de los conos alu...   
...     ...          ...                                                ...   
8206   8207   4-Dec-2020  Elaboración de un manual de procedimientos e i...   
8207   8208   8-Mar-2021  Condiciones de vida de los estudiantes de Geoq...   
8208   8209   8-Mar-2021  Análisis de los elementos pospornográficos de ...   
8209   8210  11-Mar-2021  La Fecundidad Adolescente en Venezuela vista a...   
8210   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
1                                     Rivera, Elizabeth       2 MB   
2     Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  995.92 kB   
3                                    D'Amico, Walter F.  506.32 kB   
4                  Cragno, Analia V.; Romero, Marcos A.    3.02 MB   
...                                                 ...        ...   
8206                               Armas R., Rodolfo J.  924.37 kB   
8207                     Pérez, Andrea; Pinto, Hendrith    7.21 MB   
8208                                   Belandria, Sofía    1.13 MB   
8209                              Simao, Edith Carolina    3.56 MB   
8210                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1     http://saber.ucv.ve/handle/10872/16245   
2        http://saber.ucv.ve/handle/10872/29   
3        http://saber.ucv.ve/handle/10872/32   
4        http://saber.ucv.ve/handle/10872/31   
...                                      ...   
8206  http://saber.ucv.ve/handle/10872/20906   
8207  http://saber.ucv.ve/handle/10872/20969   
8208  http://saber.ucv.ve/handle/10872/20970   
8209  http://saber.ucv.ve/handle/10872/20973   
8210  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1     http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN   
2     http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN   
3     http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN   
4     http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN   
...                                                 ...      ...   
8206  http://saber.ucv.ve/bitstream/10872/20906/1/TE...      NaN   
8207  http://saber.ucv.ve/bitstream/10872/20969/1/CO...      NaN   
8208  http://saber.ucv.ve/bitstream/10872/20970/1/So...      NaN   
8209  http://saber.ucv.ve/bitstream/10872/20973/1/TF...      NaN   
8210  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  
0        thesis_pdf_all/1.pdf   True  
1        thesis_pdf_all/2.pdf   True  
2        thesis_pdf_all/3.pdf   True  
3        thesis_pdf_all/4.pdf   True  
4        thesis_pdf_all/5.pdf   True  
...                       ...    ...  
8206  thesis_pdf_all/8207.pdf   True  
8207  thesis_pdf_all/8208.pdf   True  
8208  thesis_pdf_all/8209.pdf   True  
8209  thesis_pdf_all/8210.pdf   True  
8210  thesis_pdf_all/8211.pdf   True  

[8211 rows x 10 columns]

In [30]:
## Data frame con los que existe
df_exist = df[df['exist'] == True]
df_exist

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1         2         2010  La Economía del Ocio como una alternativa de e...   
2         3  27-May-2011  La producción de conocimiento : un análisis mé...   
3         4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4         5  27-May-2011  Evaluación geofísica integral de los conos alu...   
...     ...          ...                                                ...   
8206   8207   4-Dec-2020  Elaboración de un manual de procedimientos e i...   
8207   8208   8-Mar-2021  Condiciones de vida de los estudiantes de Geoq...   
8208   8209   8-Mar-2021  Análisis de los elementos pospornográficos de ...   
8209   8210  11-Mar-2021  La Fecundidad Adolescente en Venezuela vista a...   
8210   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
1                                     Rivera, Elizabeth       2 MB   
2     Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  995.92 kB   
3                                    D'Amico, Walter F.  506.32 kB   
4                  Cragno, Analia V.; Romero, Marcos A.    3.02 MB   
...                                                 ...        ...   
8206                               Armas R., Rodolfo J.  924.37 kB   
8207                     Pérez, Andrea; Pinto, Hendrith    7.21 MB   
8208                                   Belandria, Sofía    1.13 MB   
8209                              Simao, Edith Carolina    3.56 MB   
8210                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1     http://saber.ucv.ve/handle/10872/16245   
2        http://saber.ucv.ve/handle/10872/29   
3        http://saber.ucv.ve/handle/10872/32   
4        http://saber.ucv.ve/handle/10872/31   
...                                      ...   
8206  http://saber.ucv.ve/handle/10872/20906   
8207  http://saber.ucv.ve/handle/10872/20969   
8208  http://saber.ucv.ve/handle/10872/20970   
8209  http://saber.ucv.ve/handle/10872/20973   
8210  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1     http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN   
2     http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN   
3     http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN   
4     http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN   
...                                                 ...      ...   
8206  http://saber.ucv.ve/bitstream/10872/20906/1/TE...      NaN   
8207  http://saber.ucv.ve/bitstream/10872/20969/1/CO...      NaN   
8208  http://saber.ucv.ve/bitstream/10872/20970/1/So...      NaN   
8209  http://saber.ucv.ve/bitstream/10872/20973/1/TF...      NaN   
8210  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  
0        thesis_pdf_all/1.pdf   True  
1        thesis_pdf_all/2.pdf   True  
2        thesis_pdf_all/3.pdf   True  
3        thesis_pdf_all/4.pdf   True  
4        thesis_pdf_all/5.pdf   True  
...                       ...    ...  
8206  thesis_pdf_all/8207.pdf   True  
8207  thesis_pdf_all/8208.pdf   True  
8208  thesis_pdf_all/8209.pdf   True  
8209  thesis_pdf_all/8210.pdf   True  
8210  thesis_pdf_all/8211.pdf   True  

[8173 rows x 10 columns]

In [34]:
len(df) - len(df_exist)

38

In [159]:
(38*100)/8211

0.4627938131774449

### Archivos no descargados

De todas las tesis inicialmente mapeadas 8211, 38 no fueron descargadas, esto se debe a que hubo varios erroes en la descarga. Dando un total de 8173 Tesis de pregrado que fueron descargadas o completamente o parcialmente. 
Esto significa 0.462% de perdidas en la descarga

Este comentario parcial se debe a que incluso si fueron descargadas y el archivo existe estas tesis pudieron descargarse con errores, volviendolas inaccesibles.

Para esto se realiza tener un data frame que no enseñe solo las tesis que son accesibles, incluso si son imagenes.

Este proceso se realiza en el notebook `notebooks/delete_not_exist.ipynb` Donde los archivos descargados que tienen errores son eliminados, dando como resultado el segundo dataframe con solo los archivos pdfs que pueden ser realmente leidos. Este archivo es: `data/url_thesis_8211_with_pdf_scan.csv`

In [35]:
## Muestra el archivo con los archivos descargados en una nueva columna
import pandas as pd
import numpy as np

df = pd.read_csv("data/url_thesis_8211_with_pdf_scan.csv")
df

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1         2         2010  La Economía del Ocio como una alternativa de e...   
2         3  27-May-2011  La producción de conocimiento : un análisis mé...   
3         4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4         5  27-May-2011  Evaluación geofísica integral de los conos alu...   
...     ...          ...                                                ...   
8168   8207   4-Dec-2020  Elaboración de un manual de procedimientos e i...   
8169   8208   8-Mar-2021  Condiciones de vida de los estudiantes de Geoq...   
8170   8209   8-Mar-2021  Análisis de los elementos pospornográficos de ...   
8171   8210  11-Mar-2021  La Fecundidad Adolescente en Venezuela vista a...   
8172   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
1                                     Rivera, Elizabeth       2 MB   
2     Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  995.92 kB   
3                                    D'Amico, Walter F.  506.32 kB   
4                  Cragno, Analia V.; Romero, Marcos A.    3.02 MB   
...                                                 ...        ...   
8168                               Armas R., Rodolfo J.  924.37 kB   
8169                     Pérez, Andrea; Pinto, Hendrith    7.21 MB   
8170                                   Belandria, Sofía    1.13 MB   
8171                              Simao, Edith Carolina    3.56 MB   
8172                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1     http://saber.ucv.ve/handle/10872/16245   
2        http://saber.ucv.ve/handle/10872/29   
3        http://saber.ucv.ve/handle/10872/32   
4        http://saber.ucv.ve/handle/10872/31   
...                                      ...   
8168  http://saber.ucv.ve/handle/10872/20906   
8169  http://saber.ucv.ve/handle/10872/20969   
8170  http://saber.ucv.ve/handle/10872/20970   
8171  http://saber.ucv.ve/handle/10872/20973   
8172  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1     http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN   
2     http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN   
3     http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN   
4     http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN   
...                                                 ...      ...   
8168  http://saber.ucv.ve/bitstream/10872/20906/1/TE...      NaN   
8169  http://saber.ucv.ve/bitstream/10872/20969/1/CO...      NaN   
8170  http://saber.ucv.ve/bitstream/10872/20970/1/So...      NaN   
8171  http://saber.ucv.ve/bitstream/10872/20973/1/TF...      NaN   
8172  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  
0        thesis_pdf_all/1.pdf   True   True  
1        thesis_pdf_all/2.pdf   True   True  
2        thesis_pdf_all/3.pdf   True   True  
3        thesis_pdf_all/4.pdf   True   True  
4        thesis_pdf_all/5.pdf   True   True  
...                       ...    ...    ...  
8168  thesis_pdf_all/8207.pdf   True   True  
8169  thesis_pdf_all/8208.pdf   True   True  
8170  thesis_pdf_all/8209.pdf   True   True  
8171  thesis_pdf_all/8210.pdf   True   True  
8172  thesis_pdf_all/8211.pdf   True   True  

[8173 rows x 11 columns]

In [36]:
isPDF_df = df[df['isPDF']==True]
isPDF_df

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1         2         2010  La Economía del Ocio como una alternativa de e...   
2         3  27-May-2011  La producción de conocimiento : un análisis mé...   
3         4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4         5  27-May-2011  Evaluación geofísica integral de los conos alu...   
...     ...          ...                                                ...   
8168   8207   4-Dec-2020  Elaboración de un manual de procedimientos e i...   
8169   8208   8-Mar-2021  Condiciones de vida de los estudiantes de Geoq...   
8170   8209   8-Mar-2021  Análisis de los elementos pospornográficos de ...   
8171   8210  11-Mar-2021  La Fecundidad Adolescente en Venezuela vista a...   
8172   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
1                                     Rivera, Elizabeth       2 MB   
2     Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  995.92 kB   
3                                    D'Amico, Walter F.  506.32 kB   
4                  Cragno, Analia V.; Romero, Marcos A.    3.02 MB   
...                                                 ...        ...   
8168                               Armas R., Rodolfo J.  924.37 kB   
8169                     Pérez, Andrea; Pinto, Hendrith    7.21 MB   
8170                                   Belandria, Sofía    1.13 MB   
8171                              Simao, Edith Carolina    3.56 MB   
8172                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1     http://saber.ucv.ve/handle/10872/16245   
2        http://saber.ucv.ve/handle/10872/29   
3        http://saber.ucv.ve/handle/10872/32   
4        http://saber.ucv.ve/handle/10872/31   
...                                      ...   
8168  http://saber.ucv.ve/handle/10872/20906   
8169  http://saber.ucv.ve/handle/10872/20969   
8170  http://saber.ucv.ve/handle/10872/20970   
8171  http://saber.ucv.ve/handle/10872/20973   
8172  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1     http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN   
2     http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN   
3     http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN   
4     http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN   
...                                                 ...      ...   
8168  http://saber.ucv.ve/bitstream/10872/20906/1/TE...      NaN   
8169  http://saber.ucv.ve/bitstream/10872/20969/1/CO...      NaN   
8170  http://saber.ucv.ve/bitstream/10872/20970/1/So...      NaN   
8171  http://saber.ucv.ve/bitstream/10872/20973/1/TF...      NaN   
8172  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  
0        thesis_pdf_all/1.pdf   True   True  
1        thesis_pdf_all/2.pdf   True   True  
2        thesis_pdf_all/3.pdf   True   True  
3        thesis_pdf_all/4.pdf   True   True  
4        thesis_pdf_all/5.pdf   True   True  
...                       ...    ...    ...  
8168  thesis_pdf_all/8207.pdf   True   True  
8169  thesis_pdf_all/8208.pdf   True   True  
8170  thesis_pdf_all/8209.pdf   True   True  
8171  thesis_pdf_all/8210.pdf   True   True  
8172  thesis_pdf_all/8211.pdf   True   True  

[7883 rows x 11 columns]

In [37]:
len(df) - len(isPDF_df)

290

In [160]:
(290*100)/8173

3.5482686895876667

### Tercer Dataframe: Archivos con errores

De las 8211 tesis 8173 se lograron descargar y de las 8173 tesis 290 se descargaron con error (lo que quiere decir que 3.54% de las tesis descargadas tienen error), errores que la hacen imposible acceder a ella, ni como imagen ni como texto. Dando como total 7883 tesis de pregrado que si se pueden leer como pdf, es decir, son archivos que si se pueden abrir con un visor o editor de pdfs.

### Es imagen o es texto

De estas 7883 tesis que vamos a tomar, tambien en el archivo `notebooks/delete_not_exist.ipynb` Podemos encontrar las 5 condiciones de saber si el archivo es imagen o es pdf.

## Criterio de clasificación

Para determinar que una tesis es escaneada y no contiene texto, se revizaron 5 tesis del primer lote inicial y se determino revisando todas las hojas manualmente que la mayoria eran imagenes. Con esto en mente y teniendo en cuenta que se requiere clasificar por carrera y esta informacion se encuentra normalmente en las primeras paginas de las tesis, el criterio para determinar si una tesis mayoritariamente imagenes que requieres OCR (Reconocimineto optico de caracteres) para pasar a texto queda de la siguiente manera:

- La primera hoja es imagen o no.
- La segunda hoja es imagen o no.
- La hoja del medio mas uno es imagen o no.
- La hoja del final -20 es imagen o no (esto es por la bibliografia).
- Una hoja aleatoria que no sea la primera ni la segunda es imagen o no.

Para determinar si es imagen o no se utilizo la funcion de `pdfplumber` `extract_text()` si esta funcion retornaba '' un caracter vacio significaba que la pagina no contiene texto crudo si no en formato de una imagen escaneada.

Si el archivo pdf cumple con 3 de las 5 condiciones se considera que es imagen, de lo contrario es texto.

Las 5 tesis revisadas en cuestion:
1) http://saber.ucv.ve/handle/10872/315   
2) http://saber.ucv.ve/handle/10872/336   
3) http://saber.ucv.ve/handle/10872/341   
4) http://saber.ucv.ve/handle/10872/318   
5) http://saber.ucv.ve/handle/10872/314 (El resumen era texto pero el resto era escaneado) 


## Cuarto data frame: Imagen o texto?

Este dataframe contiene informacion de si la tesis es imagen o es texto `data/url_thesis_8211_with_pdf_scan_image.csv`

In [47]:
## Muestra los archivos que son imagenes en una nueva columna
import pandas as pd
import numpy as np

df = pd.read_csv("data/url_thesis_8211_with_pdf_scan_image.csv")
df

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1         2         2010  La Economía del Ocio como una alternativa de e...   
2         3  27-May-2011  La producción de conocimiento : un análisis mé...   
3         4  27-May-2011  Desarrollo de un algoritmo computacional para ...   
4         5  27-May-2011  Evaluación geofísica integral de los conos alu...   
...     ...          ...                                                ...   
7878   8207   4-Dec-2020  Elaboración de un manual de procedimientos e i...   
7879   8208   8-Mar-2021  Condiciones de vida de los estudiantes de Geoq...   
7880   8209   8-Mar-2021  Análisis de los elementos pospornográficos de ...   
7881   8210  11-Mar-2021  La Fecundidad Adolescente en Venezuela vista a...   
7882   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
1                                     Rivera, Elizabeth       2 MB   
2     Adrian Mogollón, Julia; Echezuria Perozo, Jorg...  995.92 kB   
3                                    D'Amico, Walter F.  506.32 kB   
4                  Cragno, Analia V.; Romero, Marcos A.    3.02 MB   
...                                                 ...        ...   
7878                               Armas R., Rodolfo J.  924.37 kB   
7879                     Pérez, Andrea; Pinto, Hendrith    7.21 MB   
7880                                   Belandria, Sofía    1.13 MB   
7881                              Simao, Edith Carolina    3.56 MB   
7882                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1     http://saber.ucv.ve/handle/10872/16245   
2        http://saber.ucv.ve/handle/10872/29   
3        http://saber.ucv.ve/handle/10872/32   
4        http://saber.ucv.ve/handle/10872/31   
...                                      ...   
7878  http://saber.ucv.ve/handle/10872/20906   
7879  http://saber.ucv.ve/handle/10872/20969   
7880  http://saber.ucv.ve/handle/10872/20970   
7881  http://saber.ucv.ve/handle/10872/20973   
7882  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1     http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN   
2     http://saber.ucv.ve/bitstream/10872/29/1/TL-EB...      NaN   
3     http://saber.ucv.ve/bitstream/10872/32/1/Tesis...      NaN   
4     http://saber.ucv.ve/bitstream/10872/31/1/Evalu...      NaN   
...                                                 ...      ...   
7878  http://saber.ucv.ve/bitstream/10872/20906/1/TE...      NaN   
7879  http://saber.ucv.ve/bitstream/10872/20969/1/CO...      NaN   
7880  http://saber.ucv.ve/bitstream/10872/20970/1/So...      NaN   
7881  http://saber.ucv.ve/bitstream/10872/20973/1/TF...      NaN   
7882  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  isScan  
0        thesis_pdf_all/1.pdf   True   True   False  
1        thesis_pdf_all/2.pdf   True   True   False  
2        thesis_pdf_all/3.pdf   True   True   False  
3        thesis_pdf_all/4.pdf   True   True   False  
4        thesis_pdf_all/5.pdf   True   True   False  
...                       ...    ...    ...     ...  
7878  thesis_pdf_all/8207.pdf   True   True   False  
7879  thesis_pdf_all/8208.pdf   True   True   False  
7880  thesis_pdf_all/8209.pdf   True   True   False  
7881  thesis_pdf_all/8210.pdf   True   True   False  
7882  thesis_pdf_all/8211.pdf   True   True   False  

[7883 rows x 12 columns]

In [48]:
df_scan = df[df['isScan']==False]
df_scan = df_scan.sort_values('isScan', ascending=False)
df_scan

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
5258   5539  29-Mar-2017  Diseño de manejo de aguas de mina, en el PIT b...   
5286   5573  30-Mar-2017  Identificación y cuantificación de hidrocarbur...   
5285   5572  30-Mar-2017  Evaluación de catalizadores soportados de Pd-M...   
5284   5571  30-Mar-2017  Estudio de la solubilidad para sistemas binari...   
...     ...          ...                                                ...   
2670   2925  16-Dec-2014  Identificación y clonación del gen que codific...   
2669   2924  16-Dec-2014  Estudio del desarrollo postnatal de la extensi...   
2668   2923  16-Dec-2014  Zooplancton en seis cuerpos de agua de la regi...   
2667   2922  15-Dec-2014  Distribución de la enzima 3α-hidroxiesteroide ...   
7882   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                          thesis_author      size  \
0               De Gregorio S., Luis T.   6.77 MB   
5258            Van Zanten V., Johan O.    6.7 MB   
5286  Hergueta Possamai, Verónica Yrama   3.63 MB   
5285       Tovar Flores, Manuel Enrique   2.91 MB   
5284       Rampersad Avilán, Jeniree G.   1.55 MB   
...                                 ...       ...   
2670                     Ibarra, Ana V.   3.41 MB   
2669                Pernía, Marianny J.   6.14 MB   
2668                Baptista, Esther N.   3.43 MB   
2667               Alcalá, Katherine A.  30.67 MB   
7882        Utrera Ávila, Joselin Marly   2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
5258  http://saber.ucv.ve/handle/10872/15542   
5286  http://saber.ucv.ve/handle/10872/15578   
5285  http://saber.ucv.ve/handle/10872/15579   
5284  http://saber.ucv.ve/handle/10872/15580   
...                                      ...   
2670   http://saber.ucv.ve/handle/10872/8005   
2669   http://saber.ucv.ve/handle/10872/7995   
2668   http://saber.ucv.ve/handle/10872/8006   
2667   http://saber.ucv.ve/handle/10872/7989   
7882  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
5258  http://saber.ucv.ve/bitstream/10872/15542/1/TE...      NaN   
5286  http://saber.ucv.ve/bitstream/10872/15578/1/TE...      NaN   
5285  http://saber.ucv.ve/bitstream/10872/15579/1/EV...      NaN   
5284  http://saber.ucv.ve/bitstream/10872/15580/1/te...      NaN   
...                                                 ...      ...   
2670  http://saber.ucv.ve/bitstream/10872/8005/1/Tes...      NaN   
2669  http://saber.ucv.ve/bitstream/10872/7995/1/Tes...      NaN   
2668  http://saber.ucv.ve/bitstream/10872/8006/1/Tes...      NaN   
2667  http://saber.ucv.ve/bitstream/10872/7989/1/Tes...      NaN   
7882  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  isScan  
0        thesis_pdf_all/1.pdf   True   True   False  
5258  thesis_pdf_all/5539.pdf   True   True   False  
5286  thesis_pdf_all/5573.pdf   True   True   False  
5285  thesis_pdf_all/5572.pdf   True   True   False  
5284  thesis_pdf_all/5571.pdf   True   True   False  
...                       ...    ...    ...     ...  
2670  thesis_pdf_all/2925.pdf   True   True   False  
2669  thesis_pdf_all/2924.pdf   True   True   False  
2668  thesis_pdf_all/2923.pdf   True   True   False  
2667  thesis_pdf_all/2922.pdf   True   True   False  
7882  thesis_pdf_all/8211.pdf   True   True   False  

[7801 rows x 12 columns]

In [49]:
len(df) - len(df_scan)

82

In [161]:
(82*100)/7883

1.0402131168336928

### Resultados de imagenes
Solo 82 tesis eran imagenes, el resto era texto crudo, entonces de 7883 tesis que se descargaron correctamente, 7801 eran texto. Sacando resultados parecidos entones a la hora de tamaños que nos queda. 

Lo que significa que solo 1.04% de las tesis descargadas que se pueden abrir, son imagenes.

In [51]:
df_byte_text = df_scan.copy()
df_byte_text['size'] = do_bytes(df_scan['size'])
df_byte_text

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
5258   5539  29-Mar-2017  Diseño de manejo de aguas de mina, en el PIT b...   
5286   5573  30-Mar-2017  Identificación y cuantificación de hidrocarbur...   
5285   5572  30-Mar-2017  Evaluación de catalizadores soportados de Pd-M...   
5284   5571  30-Mar-2017  Estudio de la solubilidad para sistemas binari...   
...     ...          ...                                                ...   
2670   2925  16-Dec-2014  Identificación y clonación del gen que codific...   
2669   2924  16-Dec-2014  Estudio del desarrollo postnatal de la extensi...   
2668   2923  16-Dec-2014  Zooplancton en seis cuerpos de agua de la regi...   
2667   2922  15-Dec-2014  Distribución de la enzima 3α-hidroxiesteroide ...   
7882   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                          thesis_author   size  \
0               De Gregorio S., Luis T.   6.77   
5258            Van Zanten V., Johan O.   6.70   
5286  Hergueta Possamai, Verónica Yrama   3.63   
5285       Tovar Flores, Manuel Enrique   2.91   
5284       Rampersad Avilán, Jeniree G.   1.55   
...                                 ...    ...   
2670                     Ibarra, Ana V.   3.41   
2669                Pernía, Marianny J.   6.14   
2668                Baptista, Esther N.   3.43   
2667               Alcalá, Katherine A.  30.67   
7882        Utrera Ávila, Joselin Marly   2.33   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
5258  http://saber.ucv.ve/handle/10872/15542   
5286  http://saber.ucv.ve/handle/10872/15578   
5285  http://saber.ucv.ve/handle/10872/15579   
5284  http://saber.ucv.ve/handle/10872/15580   
...                                      ...   
2670   http://saber.ucv.ve/handle/10872/8005   
2669   http://saber.ucv.ve/handle/10872/7995   
2668   http://saber.ucv.ve/handle/10872/8006   
2667   http://saber.ucv.ve/handle/10872/7989   
7882  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
5258  http://saber.ucv.ve/bitstream/10872/15542/1/TE...      NaN   
5286  http://saber.ucv.ve/bitstream/10872/15578/1/TE...      NaN   
5285  http://saber.ucv.ve/bitstream/10872/15579/1/EV...      NaN   
5284  http://saber.ucv.ve/bitstream/10872/15580/1/te...      NaN   
...                                                 ...      ...   
2670  http://saber.ucv.ve/bitstream/10872/8005/1/Tes...      NaN   
2669  http://saber.ucv.ve/bitstream/10872/7995/1/Tes...      NaN   
2668  http://saber.ucv.ve/bitstream/10872/8006/1/Tes...      NaN   
2667  http://saber.ucv.ve/bitstream/10872/7989/1/Tes...      NaN   
7882  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  isScan  
0        thesis_pdf_all/1.pdf   True   True   False  
5258  thesis_pdf_all/5539.pdf   True   True   False  
5286  thesis_pdf_all/5573.pdf   True   True   False  
5285  thesis_pdf_all/5572.pdf   True   True   False  
5284  thesis_pdf_all/5571.pdf   True   True   False  
...                       ...    ...    ...     ...  
2670  thesis_pdf_all/2925.pdf   True   True   False  
2669  thesis_pdf_all/2924.pdf   True   True   False  
2668  thesis_pdf_all/2923.pdf   True   True   False  
2667  thesis_pdf_all/2922.pdf   True   True   False  
7882  thesis_pdf_all/8211.pdf   True   True   False  

[7801 rows x 12 columns]

In [52]:
df_byte_text.describe()

index         size  resumen
count  7801.000000  7801.000000      0.0
mean   4235.204333     5.264895      NaN
std    2303.156472    14.888695      NaN
min       1.000000     0.005680      NaN
25%    2273.000000     1.350000      NaN
50%    4243.000000     2.630000      NaN
75%    6227.000000     4.990000      NaN
max    8211.000000   463.160000      NaN

In [53]:
size_GB = df_byte_text['size'].sum()/1000
size_GB

41.07144575

### Tamaño de las tesis de solo texto
De estos valores la media es 5.2 MB con una desviacion estandar de 14.8 MB donde el valor minimo es 0.00568 MB o lo que es igual a 5.68 kB. El archivo más pesado siendo de 463 MB.

El tamaño total de tesis es 41.071 GB.

## Quinto data frame: Escuelas

### Criterio de busqueda de carrera

Para determinar la escuela a la que pertenece una tesis, se tiene que pasar el texto de las paginas pdf a texto crudo, este texto tiene que pasar por un proceso de limpiado para eliminar multiples espacios en blanco, acentos y saltos de linea. Estos luego son comparados por las carreras en `escuelas.json`.


```json
"facultad": "Facultad de Ingeniería",
    "escuela": [
        "Escuela de Ingeniería Civil",
        "Escuela de Ingeniería Eléctrica",
        "Escuela de Geología, Minas y Geofísica",
        "Escuela de Ingeniería Química",
        "Escuela de Ingeniería de Petroleo",
        "Escuela de Ingeniería Mecanica",
        "Escuela de Ingeniería Metalúrgica y Ciencia de los Materiales"
    ]
    },
    {
        "facultad": "Facultad de Agronomía",
        "escuela": [
            "Escuela de Agronomía"
        ]
    },
    {
        "facultad": "Facultad de Arquitectura y Urbanismo",
        "escuela": [
            "Escuela de Arquitectura"
        ]
    },
    {
        "facultad": "Facultad de Ciencias",
        "escuela": [
            "Escuela de Biología",
            "Escuela de Computación",
            "Escuela de Física",
            "Escuela de Geoquímica",
            "Escuela de Matemática",
            "Escuela de Química"
        ]
    },
    {
        "facultad": "Facultad de Ciencias Económicas y Sociales",
        "escuela":[
            "Escuela de Administración y Contaduría",
            "Escuela de Antropología",
            "Escuela de Estadística y Ciencias Actuariales",
            "Escuela de Economía",
            "Escuela de Estudios Internacionales",
            "Escuela de Sociología",
            "Escuela de Trabajo Social"
        ] 
    },

```

No es sensible a mayusculas.
La oracion de especificacion de la escuela tiene que ser exacta, esta se puede ser encontrar en el encabezado o por la firma del acta.
Por lo tanto se especifica en que paginas se buscara la escuela en las primeras 10 paginas si ocurre 1 match entonces se declarara esa como la escuela.

Se utilizo apache tika para obtener el texto de las carreras, el hecho de utilizar apache tika en vez de pdfplumber, es que apache tika es considerablemente mejor al tratar con acentos y pdfs, es en general mas preciso obteniendo las carreras.

Solo se comprobo en las primeras 6000 palabras y las ultimas 6000 paginas si habia un encuentro con una carrera. Esto ayudo a que el procesado de cada archivo fuera mas rapido, debido a que si se trata de procesar toda la tesis en apache tika esto daba errores.

La memoria y procesamiento del matcher tuvo que hacerse por batches de 20 con cada uno teniendo al alrededor de 391 tesis. Despues cada uno de los vectores resultados se unieron para dar el dataframe final con todas las tesis.

No se tiene una metrica exacta de cual fue la precision de carreras clasificadas apropiadamente, pero realizando revisiones rapidas se nota que la mayoria fue clasificada correctamente. 

De **7801** se paso a **7149** Esta es la cantidad de tesis con escuela encontrable, es decir que no dio no School o None.

In [76]:
import json
  
# Opening JSON file
f = open('data/escuelas.json')
  
# returns JSON object as 
# a dictionary
escuelas = json.load(f)

escuelas

[{'facultad': 'Facultad de Ingeniería',
  'escuela': ['Escuela de Ingeniería Civil',
   'Escuela de Ingeniería Eléctrica',
   'Escuela de Geología, Minas y Geofísica',
   'Escuela de Ingeniería Química',
   'Escuela de Ingeniería de Petroleo',
   'Escuela de Ingeniería Mecanica',
   'Escuela de Ingeniería Metalúrgica y Ciencia de los Materiales']},
 {'facultad': 'Facultad de Agronomía', 'escuela': ['Escuela de Agronomía']},
 {'facultad': 'Facultad de Arquitectura y Urbanismo',
  'escuela': ['Escuela de Arquitectura']},
 {'facultad': 'Facultad de Ciencias',
  'escuela': ['Escuela de Biología',
   'Escuela de Computación',
   'Escuela de Física',
   'Escuela de Geoquímica',
   'Escuela de Matemática',
   'Escuela de Química']},
 {'facultad': 'Facultad de Ciencias Económicas y Sociales',
  'escuela': ['Escuela de Administración y Contaduría',
   'Escuela de Antropología',
   'Escuela de Estadística y Ciencias Actuariales',
   'Escuela de Economía',
   'Escuela de Estudios Internacionales'

In [84]:
counter = 0
for i in escuelas:
    counter += 1
    
counter
# Cantida de facultades

11

In [85]:
counter = 0
for i in escuelas:
    for esc in i['escuela']:
        counter += 1
        
counter
#Cantidad de escuelas

43

In [54]:
## Muestra los archivos que son imagenes en una nueva columna
import pandas as pd
import numpy as np

df = pd.read_csv("data/url_thesis_8211_with_pdf_scan_image.csv")
df_scan = df[df['isScan']==False]
df_scan = df_scan.sort_values('isScan', ascending=False)
df_scan

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
5258   5539  29-Mar-2017  Diseño de manejo de aguas de mina, en el PIT b...   
5286   5573  30-Mar-2017  Identificación y cuantificación de hidrocarbur...   
5285   5572  30-Mar-2017  Evaluación de catalizadores soportados de Pd-M...   
5284   5571  30-Mar-2017  Estudio de la solubilidad para sistemas binari...   
...     ...          ...                                                ...   
2670   2925  16-Dec-2014  Identificación y clonación del gen que codific...   
2669   2924  16-Dec-2014  Estudio del desarrollo postnatal de la extensi...   
2668   2923  16-Dec-2014  Zooplancton en seis cuerpos de agua de la regi...   
2667   2922  15-Dec-2014  Distribución de la enzima 3α-hidroxiesteroide ...   
7882   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                          thesis_author      size  \
0               De Gregorio S., Luis T.   6.77 MB   
5258            Van Zanten V., Johan O.    6.7 MB   
5286  Hergueta Possamai, Verónica Yrama   3.63 MB   
5285       Tovar Flores, Manuel Enrique   2.91 MB   
5284       Rampersad Avilán, Jeniree G.   1.55 MB   
...                                 ...       ...   
2670                     Ibarra, Ana V.   3.41 MB   
2669                Pernía, Marianny J.   6.14 MB   
2668                Baptista, Esther N.   3.43 MB   
2667               Alcalá, Katherine A.  30.67 MB   
7882        Utrera Ávila, Joselin Marly   2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
5258  http://saber.ucv.ve/handle/10872/15542   
5286  http://saber.ucv.ve/handle/10872/15578   
5285  http://saber.ucv.ve/handle/10872/15579   
5284  http://saber.ucv.ve/handle/10872/15580   
...                                      ...   
2670   http://saber.ucv.ve/handle/10872/8005   
2669   http://saber.ucv.ve/handle/10872/7995   
2668   http://saber.ucv.ve/handle/10872/8006   
2667   http://saber.ucv.ve/handle/10872/7989   
7882  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
5258  http://saber.ucv.ve/bitstream/10872/15542/1/TE...      NaN   
5286  http://saber.ucv.ve/bitstream/10872/15578/1/TE...      NaN   
5285  http://saber.ucv.ve/bitstream/10872/15579/1/EV...      NaN   
5284  http://saber.ucv.ve/bitstream/10872/15580/1/te...      NaN   
...                                                 ...      ...   
2670  http://saber.ucv.ve/bitstream/10872/8005/1/Tes...      NaN   
2669  http://saber.ucv.ve/bitstream/10872/7995/1/Tes...      NaN   
2668  http://saber.ucv.ve/bitstream/10872/8006/1/Tes...      NaN   
2667  http://saber.ucv.ve/bitstream/10872/7989/1/Tes...      NaN   
7882  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  isScan  
0        thesis_pdf_all/1.pdf   True   True   False  
5258  thesis_pdf_all/5539.pdf   True   True   False  
5286  thesis_pdf_all/5573.pdf   True   True   False  
5285  thesis_pdf_all/5572.pdf   True   True   False  
5284  thesis_pdf_all/5571.pdf   True   True   False  
...                       ...    ...    ...     ...  
2670  thesis_pdf_all/2925.pdf   True   True   False  
2669  thesis_pdf_all/2924.pdf   True   True   False  
2668  thesis_pdf_all/2923.pdf   True   True   False  
2667  thesis_pdf_all/2922.pdf   True   True   False  
7882  thesis_pdf_all/8211.pdf   True   True   False  

[7801 rows x 12 columns]

In [55]:
# Se tiene que dividir en 2 partes por limitaciones de la memoria
df_splited_scan = np.array_split(df_scan, 20)

In [58]:
#Esta es la cantidad de filas que tiene cada particion 391 a 390
len(df_splited_scan[0])

391

In [59]:
#Esta son todas las particiones con las escuelas clasificadas ahora toca unirlas
import pandas as pd
import numpy as np
import os

## get dataframe with text only thesis
csv_source0 = "./data/splitted/thesis_df_splited0_with_school.csv"
csv_source1 = "./data/splitted/thesis_df_splited1_with_school.csv"
csv_source2 = "./data/splitted/thesis_df_splited2_with_school.csv"
csv_source3 = "./data/splitted/thesis_df_splited3_with_school.csv"
csv_source4 = "./data/splitted/thesis_df_splited4_with_school.csv"
csv_source5 = "./data/splitted/thesis_df_splited5_with_school.csv"
csv_source6 = "./data/splitted/thesis_df_splited6_with_school.csv"
csv_source7 = "./data/splitted/thesis_df_splited7_with_school.csv"
csv_source8 = "./data/splitted/thesis_df_splited8_with_school.csv"
csv_source9 = "./data/splitted/thesis_df_splited9_with_school.csv"
csv_source10 = "./data/splitted/thesis_df_splited10_with_school.csv"
csv_source11 = "./data/splitted/thesis_df_splited11_with_school.csv"
csv_source12 = "./data/splitted/thesis_df_splited12_with_school.csv"
csv_source13 = "./data/splitted/thesis_df_splited13_with_school.csv"
csv_source14 = "./data/splitted/thesis_df_splited14_with_school.csv"
csv_source15 = "./data/splitted/thesis_df_splited15_with_school.csv"
csv_source16 = "./data/splitted/thesis_df_splited16_with_school.csv"
csv_source17 = "./data/splitted/thesis_df_splited17_with_school.csv"
csv_source18 = "./data/splitted/thesis_df_splited18_with_school.csv"
csv_source19 = "./data/splitted/thesis_df_splited19_with_school.csv"

# Unir las tesis
csv_a = []
csv_a.append(pd.read_csv(csv_source0))
csv_a.append(pd.read_csv(csv_source1))
csv_a.append(pd.read_csv(csv_source2))
csv_a.append(pd.read_csv(csv_source3))
csv_a.append(pd.read_csv(csv_source4))
csv_a.append(pd.read_csv(csv_source5))
csv_a.append(pd.read_csv(csv_source6))
csv_a.append(pd.read_csv(csv_source7))
csv_a.append(pd.read_csv(csv_source8))
csv_a.append(pd.read_csv(csv_source9))
csv_a.append(pd.read_csv(csv_source10))
csv_a.append(pd.read_csv(csv_source11))
csv_a.append(pd.read_csv(csv_source12))
csv_a.append(pd.read_csv(csv_source13))
csv_a.append(pd.read_csv(csv_source14))
csv_a.append(pd.read_csv(csv_source15))
csv_a.append(pd.read_csv(csv_source16))
csv_a.append(pd.read_csv(csv_source17))
csv_a.append(pd.read_csv(csv_source18))
csv_a.append(pd.read_csv(csv_source19))

In [66]:
concat_df = pd.concat(csv_a,axis=0)
concat_df

index  thesis_year                                       thesis_title  \
0        1         2009  Estudio de sobretensiones por maniobra debido ...   
1     5539  29-Mar-2017  Diseño de manejo de aguas de mina, en el PIT b...   
2     5573  30-Mar-2017  Identificación y cuantificación de hidrocarbur...   
3     5572  30-Mar-2017  Evaluación de catalizadores soportados de Pd-M...   
4     5571  30-Mar-2017  Estudio de la solubilidad para sistemas binari...   
..     ...          ...                                                ...   
385   2925  16-Dec-2014  Identificación y clonación del gen que codific...   
386   2924  16-Dec-2014  Estudio del desarrollo postnatal de la extensi...   
387   2923  16-Dec-2014  Zooplancton en seis cuerpos de agua de la regi...   
388   2922  15-Dec-2014  Distribución de la enzima 3α-hidroxiesteroide ...   
389   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                         thesis_author      size  \
0              De Gregorio S., Luis T.   6.77 MB   
1              Van Zanten V., Johan O.    6.7 MB   
2    Hergueta Possamai, Verónica Yrama   3.63 MB   
3         Tovar Flores, Manuel Enrique   2.91 MB   
4         Rampersad Avilán, Jeniree G.   1.55 MB   
..                                 ...       ...   
385                     Ibarra, Ana V.   3.41 MB   
386                Pernía, Marianny J.   6.14 MB   
387                Baptista, Esther N.   3.43 MB   
388               Alcalá, Katherine A.  30.67 MB   
389        Utrera Ávila, Joselin Marly   2.33 MB   

                                thesis_link  \
0    http://saber.ucv.ve/handle/10872/17712   
1    http://saber.ucv.ve/handle/10872/15542   
2    http://saber.ucv.ve/handle/10872/15578   
3    http://saber.ucv.ve/handle/10872/15579   
4    http://saber.ucv.ve/handle/10872/15580   
..                                      ...   
385   http://saber.ucv.ve/handle/10872/8005   
386   http://saber.ucv.ve/handle/10872/7995   
387   http://saber.ucv.ve/handle/10872/8006   
388   http://saber.ucv.ve/handle/10872/7989   
389  http://saber.ucv.ve/handle/10872/20974   

                                              pdf_link  resumen  \
0    http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1    http://saber.ucv.ve/bitstream/10872/15542/1/TE...      NaN   
2    http://saber.ucv.ve/bitstream/10872/15578/1/TE...      NaN   
3    http://saber.ucv.ve/bitstream/10872/15579/1/EV...      NaN   
4    http://saber.ucv.ve/bitstream/10872/15580/1/te...      NaN   
..                                                 ...      ...   
385  http://saber.ucv.ve/bitstream/10872/8005/1/Tes...      NaN   
386  http://saber.ucv.ve/bitstream/10872/7995/1/Tes...      NaN   
387  http://saber.ucv.ve/bitstream/10872/8006/1/Tes...      NaN   
388  http://saber.ucv.ve/bitstream/10872/7989/1/Tes...      NaN   
389  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                        path  exist  isPDF  isScan  \
0       thesis_pdf_all/1.pdf   True   True   False   
1    thesis_pdf_all/5539.pdf   True   True   False   
2    thesis_pdf_all/5573.pdf   True   True   False   
3    thesis_pdf_all/5572.pdf   True   True   False   
4    thesis_pdf_all/5571.pdf   True   True   False   
..                       ...    ...    ...     ...   
385  thesis_pdf_all/2925.pdf   True   True   False   
386  thesis_pdf_all/2924.pdf   True   True   False   
387  thesis_pdf_all/2923.pdf   True   True   False   
388  thesis_pdf_all/2922.pdf   True   True   False   
389  thesis_pdf_all/8211.pdf   True   True   False   

                              school_simple  
0           escuela de ingenieria electrica  
1    escuela de geologia, minas y geofisica  
2                        escuela de quimica  
3                        escuela de quimica  
4                        escuela de quimica  
..                                      ...  
385                               No School  
386                     escuela de biologia  
387       

In [92]:
concat_df_gb = concat_df.groupby('school_simple')
concat_df_gb.size()

school_simple
No School                                                        636
None                                                              16
escuela de administracion y contaduria                             1
escuela de agronomia                                              39
escuela de antropologia                                           55
escuela de artes                                                 223
escuela de bibliotecologia y archivologia                        288
escuela de bioanalisis                                             3
escuela de biologia                                              403
escuela de computacion                                           524
escuela de comunicacion social                                   520
escuela de derecho                                                 1
escuela de economia                                                1
escuela de educacion                                             444
escuela de enfermeri

In [97]:
# Lista de tesis que dieron None
gb = concat_df_gb.groups
key_list_from_gb = ['None']
for key, item in concat_df_gb:
    if key in key_list_from_gb: 
        print(concat_df_gb.get_group(key), "\n\n")

     index  thesis_year                                       thesis_title  \
312   5257  15-Mar-2017  Producción fotocatalítica de hidrógeno emplean...   
260   4515  17-Mar-2016  Optimización de las metodologías de extracción...   
51    4334  25-Feb-2016  Determinar, mediante interpretación de registr...   
72    4313  24-Feb-2016  Evaluación de impacto del cambio climático sob...   
90    4356  25-Feb-2016  Evaluación hidrogeoquímica de las aguas formac...   
100   4346  25-Feb-2016  Diseño y construcción de un modelo de detector...   
184   4445  15-Mar-2016  Dinámica de macronutrientes (C, N, Mg, P, K, y...   
185   4444  15-Mar-2016  Caracterización de hidrocarburos generados art...   
187   4442  14-Mar-2016  Estudio geoquímico de hidrocarburos mediante t...   
233   4426  14-Mar-2016  Relaciones entre atributos biológicos y suscep...   
54    5202  13-Mar-2017  Automatización del proceso de egreso del siste...   
89    5139   7-Mar-2017  Humedales costeros en el estado Falcón 

In [98]:
# Tesis que dieron No School
gb = concat_df_gb.groups
key_list_from_gb = ['No School']
for key, item in concat_df_gb:
    if key in key_list_from_gb: 
        print(concat_df_gb.get_group(key), "\n\n")

     index  thesis_year                                       thesis_title  \
15    5556  29-Mar-2017  Aplicabilidad de áridos reciclados para fabric...   
16    5555  29-Mar-2017  Hidroformilación de 1-octeno catalizada por el...   
23    5546  29-Mar-2017  Desarrollo de una aplicación web móvil multipl...   
59    5540  29-Mar-2017  Determinación de los requerimientos de equipos...   
75    5494  28-Mar-2017  Propuesta de plan de explotación quinquenal 20...   
..     ...          ...                                                ...   
205   2862   1-Dec-2014  Yo, nosotros y ellos 2.0: comparación entre la...   
226   2780  10-Nov-2014  Sistema de Información para la Planificación d...   
228   2778  10-Nov-2014  Mejoras al Prototipo de Sistema de Información...   
315   2996   2-Mar-2015  Descripción ultraestructural y molecular del p...   
385   2925  16-Dec-2014  Identificación y clonación del gen que codific...   

                           thesis_author       size  \
15      

### Cantida de escuelas con documentos

En la UCV hay un total de 43 Escuelas en 11 facultades. El gruppo None se refiere a que hubo un error al buscar la escuela en el archivo, esto puede darse porque el archivo contiene menos de 6000 palabras, esto puede haber pasado porque de los archivos clasificados como texto estos pasaron por error. Ejemplos se pueden ver en los archivos [5257, 4515, 4334]

Luego las tesis clasificadas en `No school` son las tesis donde en las primeras y ultimas 6000 no habia una compracion 1 a 1 de las escuelas.

Esto nos deja con un total de 636 escuela `No school` y 16 `None` Lo cual son 652 tesis o archivos, que no tienen escuela. Esto nos da un total de 7149 tesis que si tiene escuelas que si tienen.

Lo cual se refiere que de 7801 tesis de texto solo 7149 tiene escuela es decir se tiene una perdida de 8.35%.

### Datos interesantes

La mayor cantida de tesis son de la escuela de enfermeria con 699, seguido de la escuela de quimica con 622, la escuela de computacion con 524.

In [99]:
## get dataframe with only schools tag thesis
csv_source = "./data/thesis_7801_with_school.csv"
df_schools = pd.read_csv(csv_source)
df_schools = df_schools[df_schools['isScan']==False]
df_schools = df_schools.sort_values('isScan', ascending=False)

In [100]:
len(concat_df) - len(df_schools)

652

In [162]:
(652*100)/7801

8.357902832970131

In [102]:
df_schools

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
4762    249  13-Oct-2011  Análisis y evaluación de la aplicación de tecn...   
4773    197  30-Sep-2011  Modelo de participación comunitaria del comité...   
4772    198  30-Sep-2011  Calidad del cuidado brindado por el profesiona...   
4771    199  30-Sep-2011  Programa de educación para la prevención del v...   
...     ...          ...                                                ...   
2380   7733  30-Jul-2019  Evaluación y selección de sensores en sistemas...   
2379   7734  30-Jul-2019  Diseño de una red industrial MODBUS RTU inalám...   
2378   7735  31-Jul-2019  Estimación de la demanda y energía eléctrica e...   
2377   7736  31-Jul-2019  Diseño de un sistema de comunicación entre un ...   
7148   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
4762                             González P., Héctor V.    2.61 MB   
4773           Carrillo C., Nancy; Márquez S., Migdalia  280.68 kB   
4772     Casaña, Yurima; Florido, Lisbeth; Nieves, Rosa  481.18 kB   
4771  Colmenares, Suleimy; Contreras, Rosa Y.; Valer...  483.44 kB   
...                                                 ...        ...   
2380                            Sandia I., Alexandra C.     1.4 MB   
2379                          Temprano G., Alejandro J.    1.35 MB   
2378                                Landaeta R., Ana G.    2.45 MB   
2377                              Borges F., Gabriel A.    4.76 MB   
7148                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
4762    http://saber.ucv.ve/handle/10872/426   
4773    http://saber.ucv.ve/handle/10872/325   
4772    http://saber.ucv.ve/handle/10872/326   
4771    http://saber.ucv.ve/handle/10872/327   
...                                      ...   
2380  http://saber.ucv.ve/handle/10872/20088   
2379  http://saber.ucv.ve/handle/10872/20089   
2378  http://saber.ucv.ve/handle/10872/20090   
2377  http://saber.ucv.ve/handle/10872/20091   
7148  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
4762  http://saber.ucv.ve/bitstream/10872/426/1/TRAB...      NaN   
4773  http://saber.ucv.ve/bitstream/10872/325/1/MODE...      NaN   
4772  http://saber.ucv.ve/bitstream/10872/326/1/tesi...      NaN   
4771  http://saber.ucv.ve/bitstream/10872/327/1/TESI...      NaN   
...                                                 ...      ...   
2380  http://saber.ucv.ve/bitstream/10872/20088/1/TE...      NaN   
2379  http://saber.ucv.ve/bitstream/10872/20089/1/TE...      NaN   
2378  http://saber.ucv.ve/bitstream/10872/20090/1/La...      NaN   
2377  http://saber.ucv.ve/bitstream/10872/20091/1/TE...      NaN   
7148  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  isScan  \
0        thesis_pdf_all/1.pdf   True   True   False   
4762   thesis_pdf_all/249.pdf   True   True   False   
4773   thesis_pdf_all/197.pdf   True   True   False   
4772   thesis_pdf_all/198.pdf   True   True   False   
4771   thesis_pdf_all/199.pdf   True   True   False   
...                       ...    ...    ...     ...   
2380  thesis_pdf_all/7733.pdf   True   True   False   
2379  thesis_pdf_all/7734.pdf   True   True   False   
2378  thesis_pdf_all/7735.pdf   True   True   False   
2377  thesis_pdf_all/7736.pdf   True   True   False   
7148  thesis_pdf_all/8211.pdf   True   True   False   

                          school_simple  
0       escuela de ingenieria electrica  
4762  escuela de ingenieria de petroleo  
4773              escuela de enferm

In [104]:
## Escuelas complex
df_schools_complex = pd.read_csv('./data/thesis_7149_with_resumen_school_complex.csv')
df_schools_complex

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1       249  13-Oct-2011  Análisis y evaluación de la aplicación de tecn...   
2       197  30-Sep-2011  Modelo de participación comunitaria del comité...   
3       198  30-Sep-2011  Calidad del cuidado brindado por el profesiona...   
4       199  30-Sep-2011  Programa de educación para la prevención del v...   
...     ...          ...                                                ...   
7144   7733  30-Jul-2019  Evaluación y selección de sensores en sistemas...   
7145   7734  30-Jul-2019  Diseño de una red industrial MODBUS RTU inalám...   
7146   7735  31-Jul-2019  Estimación de la demanda y energía eléctrica e...   
7147   7736  31-Jul-2019  Diseño de un sistema de comunicación entre un ...   
7148   8211  12-Mar-2021  La industria musical surcoreana, recepción y a...   

                                          thesis_author       size  \
0                               De Gregorio S., Luis T.    6.77 MB   
1                                González P., Héctor V.    2.61 MB   
2              Carrillo C., Nancy; Márquez S., Migdalia  280.68 kB   
3        Casaña, Yurima; Florido, Lisbeth; Nieves, Rosa  481.18 kB   
4     Colmenares, Suleimy; Contreras, Rosa Y.; Valer...  483.44 kB   
...                                                 ...        ...   
7144                            Sandia I., Alexandra C.     1.4 MB   
7145                          Temprano G., Alejandro J.    1.35 MB   
7146                                Landaeta R., Ana G.    2.45 MB   
7147                              Borges F., Gabriel A.    4.76 MB   
7148                        Utrera Ávila, Joselin Marly    2.33 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1       http://saber.ucv.ve/handle/10872/426   
2       http://saber.ucv.ve/handle/10872/325   
3       http://saber.ucv.ve/handle/10872/326   
4       http://saber.ucv.ve/handle/10872/327   
...                                      ...   
7144  http://saber.ucv.ve/handle/10872/20088   
7145  http://saber.ucv.ve/handle/10872/20089   
7146  http://saber.ucv.ve/handle/10872/20090   
7147  http://saber.ucv.ve/handle/10872/20091   
7148  http://saber.ucv.ve/handle/10872/20974   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1     http://saber.ucv.ve/bitstream/10872/426/1/TRAB...      NaN   
2     http://saber.ucv.ve/bitstream/10872/325/1/MODE...      NaN   
3     http://saber.ucv.ve/bitstream/10872/326/1/tesi...      NaN   
4     http://saber.ucv.ve/bitstream/10872/327/1/TESI...      NaN   
...                                                 ...      ...   
7144  http://saber.ucv.ve/bitstream/10872/20088/1/TE...      NaN   
7145  http://saber.ucv.ve/bitstream/10872/20089/1/TE...      NaN   
7146  http://saber.ucv.ve/bitstream/10872/20090/1/La...      NaN   
7147  http://saber.ucv.ve/bitstream/10872/20091/1/TE...      NaN   
7148  http://saber.ucv.ve/bitstream/10872/20974/1/TF...      NaN   

                         path  exist  isPDF  isScan  \
0        thesis_pdf_all/1.pdf   True   True   False   
1      thesis_pdf_all/249.pdf   True   True   False   
2      thesis_pdf_all/197.pdf   True   True   False   
3      thesis_pdf_all/198.pdf   True   True   False   
4      thesis_pdf_all/199.pdf   True   True   False   
...                       ...    ...    ...     ...   
7144  thesis_pdf_all/7733.pdf   True   True   False   
7145  thesis_pdf_all/7734.pdf   True   True   False   
7146  thesis_pdf_all/7735.pdf   True   True   False   
7147  thesis_pdf_all/7736.pdf   True   True   False   
7148  thesis_pdf_all/8211.pdf   True   True   False   

                          school_simple                     school_complex  
0       escuela de ingenieria electrica    Escuela de Ingeniería Eléctrica  
1     e

In [155]:
df_gb_school_complex = df_schools_complex.groupby(df_schools_complex['school_complex'])
df_gb_school_complex.size()
#Cantidad de tesis por escuela

school_complex
Escuela Luis Razetti                                               1
Escuela de Administración y Contaduría                             1
Escuela de Agronomía                                              39
Escuela de Antropología                                           55
Escuela de Artes                                                 223
Escuela de Bibliotecología y Archivología                        288
Escuela de Bioanálisis                                             3
Escuela de Biología                                              403
Escuela de Computación                                           524
Escuela de Comunicación Social                                   520
Escuela de Derecho                                                 1
Escuela de Economía                                                1
Escuela de Educación                                             444
Escuela de Enfermería                                            699
Escuela de Filosofí

In [156]:
df_gb_school_complex.ngroups 
#Cantida de escuelas

34

In [120]:
import json
file = open('data/escuelas.json', "r")
file = json.load(file)
file

[{'facultad': 'Facultad de Ingeniería',
  'escuela': ['Escuela de Ingeniería Civil',
   'Escuela de Ingeniería Eléctrica',
   'Escuela de Geología, Minas y Geofísica',
   'Escuela de Ingeniería Química',
   'Escuela de Ingeniería de Petroleo',
   'Escuela de Ingeniería Mecanica',
   'Escuela de Ingeniería Metalúrgica y Ciencia de los Materiales']},
 {'facultad': 'Facultad de Agronomía', 'escuela': ['Escuela de Agronomía']},
 {'facultad': 'Facultad de Arquitectura y Urbanismo',
  'escuela': ['Escuela de Arquitectura']},
 {'facultad': 'Facultad de Ciencias',
  'escuela': ['Escuela de Biología',
   'Escuela de Computación',
   'Escuela de Física',
   'Escuela de Geoquímica',
   'Escuela de Matemática',
   'Escuela de Química']},
 {'facultad': 'Facultad de Ciencias Económicas y Sociales',
  'escuela': ['Escuela de Administración y Contaduría',
   'Escuela de Antropología',
   'Escuela de Estadística y Ciencias Actuariales',
   'Escuela de Economía',
   'Escuela de Estudios Internacionales'

In [148]:
#rows list initialization
rows = []
rows_school = []
  
# appending rows
for data in file:
    data_row = data['facultad']
    for row in data['escuela']:
        rows.append(data_row)
        rows_school.append(row)
        #rows.append(row)
        
#print(rows)
        
df_school = pd.DataFrame(columns = ['facultad', 'school_complex'])
df_school['facultad'] = rows
df_school['school_complex'] = rows_school
df_school

facultad  \
0                       Facultad de Ingeniería   
1                       Facultad de Ingeniería   
2                       Facultad de Ingeniería   
3                       Facultad de Ingeniería   
4                       Facultad de Ingeniería   
5                       Facultad de Ingeniería   
6                       Facultad de Ingeniería   
7                        Facultad de Agronomía   
8         Facultad de Arquitectura y Urbanismo   
9                         Facultad de Ciencias   
10                        Facultad de Ciencias   
11                        Facultad de Ciencias   
12                        Facultad de Ciencias   
13                        Facultad de Ciencias   
14                        Facultad de Ciencias   
15  Facultad de Ciencias Económicas y Sociales   
16  Facultad de Ciencias Económicas y Sociales   
17  Facultad de Ciencias Económicas y Sociales   
18  Facultad de Ciencias Económicas y Sociales   
19  Facultad de Ciencias Económicas y Sociales   
20  Facultad de Ciencias Económicas y Sociales   
21  Facultad de Ciencias Económicas y Sociales   
22  Facultad de Ciencias Jirídicas y Políticas   
23  Facultad de Ciencias Jirídicas y Políticas   
24           Facultad de Ciencias Veterinarias   
25                        Facultad de Farmacia   
26         Facultad de Humanidades y Educación   
27         Facultad de Humanidades y Educación   
28         Facultad de Humanidades y Educación   
29         Facultad de Humanidades y Educación   
30         Facultad de Humanidades y Educación   
31         Facultad de Humanidades y Educación   
32         Facultad de Humanidades y Educación   
33         Facultad de Humanidades y Educación   
34         Facultad de Humanidades y Educación   
35         Facultad de Humanidades y Educación   
36                        Facultad de Medicina   
37                        Facultad de Medicina   
38                        Facultad de Medicina   
39                        Facultad de Medicina   
40                        Facultad de Medicina   
41                        Facultad de Medicina   
42                     Facultad de Odontología   

                                       school_complex  
0                         Escuela de Ingeniería Civil  
1                     Escuela de Ingeniería Eléctrica  
2              Escuela de Geología, Minas y Geofísica  
3                       Escuela de Ingeniería Química  
4                   Escuela de Ingeniería de Petroleo  
5                      Escuela de Ingeniería Mecanica  
6   Escuela de Ingeniería Metalúrgica y Ciencia de...  
7                                Escuela de Agronomía  
8                             Escuela de Arquitectura  
9                                 Escuela de Biología  
10                             Escuela de Computación  
11                                  Escuela de Física  
12                              Escuela de Geoquímica  
13                              Escuela de Matemática  
14                                 Escuela de Química  
15             Escuela de Administración y Contaduría  
16                            Escuela de Antropología  
17      Escuela de Estadística y Ciencias Actuariales  
18                                Escuela de Economía  
19                Escuela de Estudios Internacionales  
20                              Escuela de Sociología  
21                          Escuela de Trabajo Social  
22                                 Escuela de Derecho  
23     Escuela de Estudios Políticos y Administrativo  
24                   Escuela de Ciencias Veterinarias  
25                                Escuela de Farmacia  
26                                   Escuela de Artes  
27          Escuela de Bibliotecología y Archivología  
28                     Escuela de Comunicación Social  
29                               Escuela de Educación  
30                               Escuela de Filosofía  
31                               Escuela de G

In [150]:
merged_inner = pd.merge(left=df_schools_complex, right=df_school, left_on='school_complex', right_on='school_complex')
merged_inner

index  thesis_year                                       thesis_title  \
0         1         2009  Estudio de sobretensiones por maniobra debido ...   
1      1160  25-Oct-2012  Simulación integrada subsuelo-superficie del p...   
2       350  20-Oct-2011  Diseño e implementación de un sistema de contr...   
3       831  29-Jun-2012  Estudio de transmisión de data codificada util...   
4       809  19-Jun-2012  Desarrollo de un sistema de control para mante...   
...     ...          ...                                                ...   
7144   3655  24-Sep-2015  Impacto de las limitaciones en el acceso al ag...   
7145   4627  14-Dec-2016  Elaboración de un manual de estilo para el dis...   
7146   4615  12-Dec-2016  El arte de editar libros con el Consejo de Des...   
7147   4617  12-Dec-2016  Ejercicio de la función jurisdiccional por los...   
7148      2         2010  La Economía del Ocio como una alternativa de e...   

                          thesis_author       size  \
0               De Gregorio S., Luis T.    6.77 MB   
1                    Fernández, Johanna    1.54 MB   
2                  Quiñones B., Luis A.    1.41 MB   
3                      Díaz S, Carmen S    6.17 MB   
4                      El Masri, Sharif    2.37 MB   
...                                 ...        ...   
7144  Brito, Gisela; Rodríguez, Yolanda  255.94 kB   
7145                 Uzcátegui, Beatriz    8.18 MB   
7146                    García, Dinelda   14.27 MB   
7147                 Tovar H., Janny M.  751.73 kB   
7148                  Rivera, Elizabeth       2 MB   

                                 thesis_link  \
0     http://saber.ucv.ve/handle/10872/17712   
1      http://saber.ucv.ve/handle/10872/1956   
2       http://saber.ucv.ve/handle/10872/543   
3      http://saber.ucv.ve/handle/10872/1379   
4      http://saber.ucv.ve/handle/10872/1328   
...                                      ...   
7144  http://saber.ucv.ve/handle/10872/11996   
7145  http://saber.ucv.ve/handle/10872/13948   
7146  http://saber.ucv.ve/handle/10872/13938   
7147  http://saber.ucv.ve/handle/10872/13934   
7148  http://saber.ucv.ve/handle/10872/16245   

                                               pdf_link  resumen  \
0     http://saber.ucv.ve/bitstream/10872/17712/1/TE...      NaN   
1     http://saber.ucv.ve/bitstream/10872/1956/1/Tes...      NaN   
2     http://saber.ucv.ve/bitstream/10872/543/1/Trab...      NaN   
3     http://saber.ucv.ve/bitstream/10872/1379/1/Tes...      NaN   
4     http://saber.ucv.ve/bitstream/10872/1328/1/TRA...      NaN   
...                                                 ...      ...   
7144  http://saber.ucv.ve/bitstream/10872/11996/1/br...      NaN   
7145  http://saber.ucv.ve/bitstream/10872/13948/1/In...      NaN   
7146  http://saber.ucv.ve/bitstream/10872/13938/1/Te...      NaN   
7147  http://saber.ucv.ve/bitstream/10872/13934/1/TE...      NaN   
7148  http://saber.ucv.ve/bitstream/10872/16245/1/TG...      NaN   

                         path  exist  isPDF  isScan  \
0        thesis_pdf_all/1.pdf   True   True   False   
1     thesis_pdf_all/1160.pdf   True   True   False   
2      thesis_pdf_all/350.pdf   True   True   False   
3      thesis_pdf_all/831.pdf   True   True   False   
4      thesis_pdf_all/809.pdf   True   True   False   
...                       ...    ...    ...     ...   
7144  thesis_pdf_all/3655.pdf   True   True   False   
7145  thesis_pdf_all/4627.pdf   True   True   False   
7146  thesis_pdf_all/4615.pdf   True   True   False   
7147  thesis_pdf_all/4617.pdf   True   True   False   
7148     thesis_pdf_all/2.pdf   True   True   False   

                        school_simple                   school_complex  \
0     escuela de ingenieria electrica  Escuela de Ingeniería Eléctrica   
1     escuela de ingenieria electrica  Escuela de Ingeniería Eléctrica   
2     escuela de ingenieria electrica  Escuela de Ingeniería Eléctrica   
3     escuela de ingenieria electrica  Escuela de Ingenie

In [152]:
merged_gb_complex_school = merged_inner.groupby(merged_inner['facultad'])
merged_gb_complex_school.size()
#Cantidad de escuelas por facultad

facultad
Facultad de Agronomía                           39
Facultad de Ciencias                          1846
Facultad de Ciencias Económicas y Sociales     205
Facultad de Ciencias Jirídicas y Políticas       1
Facultad de Humanidades y Educación           1964
Facultad de Ingeniería                        2388
Facultad de Medicina                           706
dtype: int64

### Facultad con mas tesis

La faculta de ingenieria es la que mas tesis posee de las tesis obtenidas en pregrado, seguidas por la facultad de ciencias.

Otro dato es que de las 8211 tesis de pregrado solo 7149 son usables realmente, con texto y con escuela. Esto significa que 1062 tesis se perdieron. Lo cual significa una perdida de 12.933% desde la data inicial.

De las 11 facultades solo estan reflejadas 7. De las 43 Escuelas solo estan reflejadas 34 escuelas en las tesis descargadas.

In [157]:
8211 - 7149

1062

In [158]:
(1062*100)/8211

12.933869199853854

### Conclusiones

- Primer dataframe (Todas las tesis): Con un total de 8211.
- Segundo dataframe (Descargado o no?): De todas las tesis inicialmente mapeadas 8211, 38 no fueron descargadas, esto se debe a que hubo varios erroes en la descarga. Dando un total de 8173 Tesis de pregrado que fueron descargadas o completamente o parcialmente. Esto significa 0.462% de perdidas en la descarga.
- Tercer dataframe (Accesible o no?) De las 8211 tesis 8173 se lograron descargar y de las 8173 tesis 290 se descargaron con error (lo que quiere decir que 3.54% de las tesis descargadas tienen error), errores que la hacen imposible acceder a ella, ni como imagen ni como texto. Dando como total 7883 tesis de pregrado que si se pueden leer como pdf, es decir, son archivos que si se pueden abrir con un visor o editor de pdfs.
- Cuarto dataframe (Imagen o texto?): Solo 82 tesis eran imagenes, el resto era texto crudo, entonces de 7883 tesis que se descargaron correctamente, 7801 eran texto. Sacando resultados parecidos entones a la hora de tamaños que nos queda.  Lo que significa que solo 1.04% de las tesis descargadas que se pueden abrir, son imagenes.
- Quinto dataframe (Escuelas): Esto nos deja con un total de 636 escuela `No school` y 16 `None` Lo cual son 652 tesis o archivos, que no tienen escuela. Esto nos da un total 7149 tesis, antes de 7801 que si tiene escuelas que si tienen. Lo cual se refiere que de 7801 tesis de texto solo 7149 tiene escuela es decir se tiene una perdida de 8.35%.

### Otros datos:

- De las 11 facultades solo estan reflejadas 7. De las 43 Escuelas solo estan reflejadas 34 escuelas en las tesis descargadas.
- La escuela mas comun es la `Escuela de Enfermería` con 699 que representa 9.77% de todas las ecuelas.
- La facultad mas comun es la `Facultad de Ingeneria` con 2388 tesis lo cual representa 33.403% de todas las tesis.
- La facultad de ciencias representa 25.82% con 1846 ejemplares.
- La escuela de quimica tiene 622 tesis en la facultad de ciencias la cual la hace la mayor escuela en ciencias en terminos de tesis. La cual representa 33.69% de las tesis (1846)
- La escuela de computacion con 524 ejemplares representa 28.38% de las tesis de la facultad de ciencias y 7.32% de las tesis de toda la universidad.

In [163]:
#Escuela de enfermeria porcentaje
699*100/7149

9.777591271506504

In [164]:
# Facultad de Ingeneria porcentaje
2388*100/7149

33.40327318506085

In [165]:
#Facultad de ciencias porcentaje
1846*100/7149

25.821793257798294

In [166]:
# Representacion mas amplia en ciencia (quimica)
622*100/1846

33.69447453954496

In [167]:
#Escuela de computacion
524*100/1846

28.385698808234018

In [168]:
524*100/7149

7.329696461043502